File: 144.二叉树的前序遍历.py
#
# @lc app=leetcode.cn id=144 lang=python3
#
# [144] 二叉树的前序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        stack = []
        ans = []
        while root is not None or stack:
            if root is not None:
                ans.append(root.val)
                if root.right is not None:
                    stack.append(root.right)
                root = root.left
            else:
                root = stack.pop()
        return ans
# @lc code=end

File: 144.二叉树的前序遍历.cpp
/*
 * @lc app=leetcode.cn id=144 lang=cpp
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include<vector>
#include<stack>
using namespace std;

class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ans;
        stack<TreeNode*> myStack;
        while(root != NULL || !myStack.empty()){
            if(root != NULL){
                ans.push_back(root->val);
                if(root->right != NULL){
                    myStack.push(root->right);
                }
                root = root->left;
            }else{
                root = myStack.top();
                myStack.pop();
            }
        }
        return ans;
    }
};
// @lc code=end

